### PR TITLE
mgr/cephadm: Fix OSD replacement in hosts with FQDN host name

### PR DESCRIPTION
**Pull request by jmolmo**
Resolves: https://tracker.ceph.com/issues/50805

Signed-off-by: Juan Miguel Olmo Martínez <jolmomar@redhat.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/services/osd.py [hunk 1]
```diff
@@ -94,7 +98,9 @@ def deploy_osd_daemons_for_existing_osds(self, host: str, service_name: str,
                                              replace_osd_ids: Optional[List[str]] = None) -> str:
 
         if replace_osd_ids is None:
-            replace_osd_ids = self.find_destroyed_osds().get(host, [])
+            # Note: OSD id claims in CRUSH map are linked to the bare name of
+            # the hostname. Use bare name in case of FQDN hostnames.
+            replace_osd_ids = self.find_destroyed_osds().get(host.split(".")[0], [])
```
Comment (adk3798) [inline src/pybind/mgr/cephadm/services/osd.py:103]: Should this be done for previews as well? https://github.com/ceph/ceph/blob/master/src/pybind/mgr/cephadm/services/osd.py#L221

Comment (jmolmo) [reply src/pybind/mgr/cephadm/services/osd.py:103]: Good catch Adam!!! Thank you man!!.

File: src/pybind/mgr/cephadm/services/osd.py [hunk 2]
```diff
@@ -46,8 +46,12 @@ def create_from_spec_one(host: str, drive_selection: DriveSelection) -> Optional
             if self.mgr.inventory.has_label(host, '_no_schedule'):
                 return None
 
+            # Note: OSD id claims in CRUSH map are linked to the bare name of
+            # the hostname. Use bare name in case of FQDN hostnames.
+            osd_id_claims_for_host = osd_id_claims.get(host.split(".")[0], [])
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/services/osd.py:51]: I somewhat don't really like the fact that the data structure `osd_id_claims` contains keys that are not consumable by the rest of the module. Can re move the responsibility from the consumer to the creator of `osd_id_claims`? 

```python
class OsdIdClaims:
    def __init__(self, d: dict):
        self._d = d
    def for_host(host):
        return self._d[host.split(".")[0]]
```

that would avoid code duplication and make sure that no one can get this wrong.

**Review by sebastian-philipp — APPROVED** [review]

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-05-15_10:00:36-rados-wip-kefu-testing-2021-05-15-1545-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-16_04:30:13-rados-wip-kefu-testing-2021-05-16-1043-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-05-16_15:20:21-rados-wip-kefu-testing-2021-05-16-2048-distro-basic-smithi/

failures tracked by 

- https://tracker.ceph.com/issues/50830

**Pull request merged**